lib/Serialization/DeserializeSIL.cpp

```diff
@@ -628,12 +630,29 @@ llvm::Expected<SILFunction *> SILDeserializer::readSILFunctionChecked(
   auto &cacheEntry = Funcs[FID - 1];
 
   if (cacheEntry.isFullyDeserialized() ||
-      (cacheEntry.isDeserialized() && (declarationOnly || forDebugScope))) {
+      (cacheEntry.isDeserialized() && declarationOnly)) {
     auto fn = cacheEntry.get();
 
-    if (fn->isZombie() && !forDebugScope)
-      return nullptr;
-    return fn;
+    // Functions onlyReferencedByDebugInfo (A) are a subset of functions
+    // referred to by debug scopes forDebugScope=true (B)
+
+    // I) Functions in A U B only ever get deserialized as zombies
+
+    // II) For rest of functions (B - A), there are two orders of
+    // deserialization:
+    //  i) Deserialized for debugging -> deserialized by linker/optimizer:
+    //    a) Deserialize as zombie. When referenced by linker/optimizer,
+    //    createDeclaration would resurrect function as normal
+    //  ii) Deserialized by linker/optimizer -> deserialized for debugging ->
+    //  no zombie created
+
+    if (fn->isZombie() && !forDebugScope) {
```

Comment:
@eeckstein I hope you can take a look at this again. Thanks

Reply:
lgtm